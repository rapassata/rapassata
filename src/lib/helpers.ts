import {
    AnyShapeTomato,
    ArrayShapedTomato,
    AtomShapedTomato,
    ObjectShapedTomato,
    RecordShapedTomato,
    Tomato,
    TomatoShape,
} from './tomatos';

// T => T | ?T
export type ProcessRequired<T, R> = R extends true ? T : T | undefined;

// Tomato<I, R, S> => I OR ?I based on R
type UnwrapTomato<T> = T extends Tomato<infer I, infer R, infer S> ? ProcessRequired<I, R> : T;

// Tomato<I, R, S> => S
type UnwrapTomatoShape<T> = T extends Tomato<infer T, infer R, infer S> ? S : T;

// Tomato<I, R, S> => Value (recursive)
export type TomatoToValues<T, U = UnwrapTomato<T>, S = UnwrapTomatoShape<T>> = S extends TomatoShape.Atom
    ? U
    : S extends TomatoShape.Array
    ? Array<UnwrapTomato<UnbakeTomato<U>>>
    : { [key in keyof U]: TomatoToValues<UnbakeTomato<U[key]>> };

// (AtomShapedTomato<...X> | ArrayShapedTomato<...X> | ...) => Tomato<...X, S>
export type UnbakeTomato<T> = T extends AtomShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Atom>
    : T extends ArrayShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Array>
    : T extends ObjectShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Object>
    : T extends RecordShapedTomato<infer U, infer R>
    ? Tomato<U, R, TomatoShape.Record>
    : T;

// (AtomShapedTomato<...X> | ArrayShapedTomato<...X> | ...) => Value (recursive)
export type Values<T> = TomatoToValues<UnbakeTomato<T>>;

export type Values2<T, U, S> = TomatoToValues<UnbakeTomato<T>, U, S>;
