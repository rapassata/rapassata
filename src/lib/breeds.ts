import {
    AnyShapeTomato,
    ArrayTomato,
    AtomShapedTomato,
    ObjectTomato,
    TomatoShape,
    Breeds,
    RecordTomato,
} from './tomatos';
import { FlowType } from './validate';
import { isObject } from './helpers';

const baseNode = <B extends { shape: TomatoShape.Atom }, T>(base: B) => {
    const node: AtomShapedTomato<T> = {
        ...base,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node;
};

const atomNode = <T>() => {
    const node: AtomShapedTomato<T> = baseNode({
        shape: TomatoShape.Atom,
    });
    return node;
};

export const string: AtomShapedTomato<string> = (() => {
    const node = atomNode<string>();
    return node.validate(x => typeof x === 'string', 'Not a string');
})();

export const any: AtomShapedTomato<any> = (() => {
    return atomNode<any>();
})();

export const number: AtomShapedTomato<number> = (() => {
    const node = atomNode<number>();
    return node.validate(x => typeof x === 'number', 'Not a number');
})();

export const boolean: AtomShapedTomato<boolean> = (() => {
    const node = atomNode<boolean>();
    return node.validate(x => typeof x === 'boolean', 'Not a boolean');
})();

export const array = <T extends AnyShapeTomato>(item: T): ArrayTomato<T> => {
    const node: ArrayTomato<T> = {
        item,
        shape: TomatoShape.Array,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(x => Array.isArray(x), 'Not an array');
};


export const object = <T>(structure: T): ObjectTomato<T> => {
    const node: ObjectTomato<T> = {
        structure,
        shape: TomatoShape.Object,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

export const objectOf = <T extends AnyShapeTomato>(val: T): RecordTomato<T> => {
    const node: RecordTomato<T> = {
        item: val,
        shape: TomatoShape.Record,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

export const breeds: Breeds = {
    string,
    number,
    boolean,
    object,
    // TODO:
    objectOf: objectOf as any,
    array,
    any,
};
