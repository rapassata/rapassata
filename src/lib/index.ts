import { Values } from './helpers';
import { Breeds } from './tomatos';

const { any, number, array, boolean, object, objectOf, string } = ('' as any) as Breeds;

// const foo: AtomShapedTomato<string, false> = string;

// type x = UnwrapTomato<typeof string>
const b = object({ o: string, r: string.require() });
type B = Values<typeof b>;

type A = Values<typeof string>;

// const c = object({ foo: array(number) });
const c = array(number);
type C = Values<typeof c>;

const Nest = object({
    age: number.defaultTo(18),
    user: object({
        address: object({
            street: string.validate(x => x !== ''),
            number: string.require(),
        }).require(),
        s: string.defaultTo('red'),
    }),
});
type Y = Values<typeof Nest>;

const Req = {
    data: any,
    user: {
        age: number.defaultTo(18),
        address: {
            street: string,
            number: string.require(),
        },
        colors: array(string.defaultTo('red')).defaultTo(''),
        ratedHobbies: objectOf(number).require(),
    },
};

type X = Values<typeof Req>;
