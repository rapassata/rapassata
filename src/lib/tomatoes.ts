import { TomatoToValues } from "./helpers";
import { FlowItem } from "./validate";

export const enum TomatoShape {
    Atom = 'Atom',
    Array = 'Array',
    Object = 'Object',
    Record = 'Record',
    Or = 'Or',
}

export type BakeTomato<T, R, S> = S extends TomatoShape.Atom ? AtomTomato<T, R> :
    S extends TomatoShape.Array ? ArrayTomato<T, R> :
    S extends TomatoShape.Object ? ObjectTomato<T, R> :
    S extends TomatoShape.Record ? RecordTomato<T, R> :
    S extends TomatoShape.Or ? OrTomato<T, R> :
    never;

export type UnbakeTomato<T> = T extends AtomTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Atom>
    : T extends ArrayTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Array>
    : T extends ObjectTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Object>
    : T extends RecordTomato<infer U, infer R>
    ? BaseTomato<U, R, TomatoShape.Record>
    : T;

type ReturnToShape<Y> = Y extends Array<infer I> ? TomatoShape.Array :
    Y extends object ? TomatoShape.Object : TomatoShape.Atom;

export interface BaseTomato<T, R, S extends TomatoShape> {
    shape: S;
    required: boolean;
    default?: TomatoToValues<T, R, S>;
    flow: Array<FlowItem>;

    require: () => BakeTomato<T, true, S>;
    defaultTo: (defVal: TomatoToValues<T, R, S>) => BakeTomato<T, true, S>;
    // transform: <Y>(t: (x: TomatoToValues<T, R, S>) => Y) => BakeTomato<Y, R, ReturnToShape<Y>>;
    transform: (t: (x: TomatoToValues<T, R, S>) => T) => BakeTomato<T, R, S>;
    validate: (fn: (val: TomatoToValues<T, R, S>) => Promise<boolean> | boolean, message?: string) => BakeTomato<T, R, S>;
}

type s = BakeTomato<boolean | string, true, TomatoShape.Or>

export interface AtomTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Atom> {
    shape: TomatoShape.Atom;
}
export interface ArrayTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Array> {
    shape: TomatoShape.Array;
    item: T;
}
export interface ObjectTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Object> {
    structure: T;
    shape: TomatoShape.Object;
}
export interface RecordTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Record> {
    item: T;
    shape: TomatoShape.Record;
}
export interface OrTomato<T, R = false> extends BaseTomato<T, R, TomatoShape.Or> {
    shape: TomatoShape.Or;
    fn: 'or';
}

export type Tomato<T, R = false> = AtomTomato<T, R> | ArrayTomato<T, R> | ObjectTomato<T, R> | RecordTomato<T, R> | OrTomato<T, R>;

// Builders
export interface Breeds {
    string: AtomTomato<string>;
    number: AtomTomato<number>;
    boolean: AtomTomato<boolean>;
    any: AtomTomato<any>;
    array: <T extends Tomato<any, any>>(x: T) => ArrayTomato<T>;
    objectOf: <V extends Tomato<any, any>>(val: V) => RecordTomato<Record<any, V>>;
    object: <V>(val: V) => ObjectTomato<V>;
}

const x = '' as any as AtomTomato<string>;
x.transform(x =>)
x.validate(x =>)