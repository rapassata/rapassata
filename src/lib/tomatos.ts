import { Values, Values2 } from './helpers';
import { FlowItem } from './validate';

export const enum TomatoShape {
    Atom = 'Atom',
    Array = 'Array',
    Object = 'Object',
    Record = 'Record',
}

export interface Tomato<
T,
R extends boolean = false,
S extends TomatoShape = TomatoShape.Atom,
BakedTomato = S extends TomatoShape.Atom ? AtomShapedTomato<T, R>
: S extends TomatoShape.Array ? ArrayShapedTomato<T, R>
: S extends TomatoShape.Object ? ObjectShapedTomato<T, R>
: S extends TomatoShape.Record ? RecordShapedTomato<T, R>
: T,
TrueBakedTomato = S extends TomatoShape.Atom ? AtomShapedTomato<T, true>
: S extends TomatoShape.Array ? ArrayShapedTomato<T, true>
: S extends TomatoShape.Object ? ObjectShapedTomato<T, true>
: S extends TomatoShape.Record ? RecordShapedTomato<T, true>
: T,
> {
    shape: TomatoShape;
    required: boolean;
    default?: Values2<T, T, S>;
    flow: Array<FlowItem>;

    // modeling methods
    require: () => TrueBakedTomato;
    defaultTo: (defVal: Values2<T, T, S>) => TrueBakedTomato;
    validate: (fn: (val: Values2<T, T, S>) => boolean | Promise<boolean>, message?: string) => BakedTomato;
}


// Tomato variances
export interface AtomShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Atom> {
    shape: TomatoShape.Atom;
}

export interface ArrayShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Array> {
    shape: TomatoShape.Array;
    item: AnyShapeTomato;
}

export interface ObjectShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Object> {
    shape: TomatoShape.Object;
    structure: T;
}

export interface RecordShapedTomato<T, R extends boolean = false> extends Tomato<T, R, TomatoShape.Record> {
    shape: TomatoShape.Record;
    structure: Record<any, T>;
}

export type AnyShapeTomato<T = any, R extends boolean = any> =
    | AtomShapedTomato<T, R>
    | ArrayShapedTomato<T, R>
    | ObjectShapedTomato<T, R>
    | RecordShapedTomato<T, R>;

type BakeShape<T> = T extends Tomato<infer U, infer R, TomatoShape.Atom> ? AtomShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Array> ? ArrayShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Object> ? ObjectShapedTomato<U, R>
: T extends Tomato<infer U, infer R, TomatoShape.Record> ? RecordShapedTomato<U, R>
: T

export type ArrayTomato<T extends AnyShapeTomato = any> = ArrayShapedTomato<T, false>;
export type ObjectTomato<T = any> = ObjectShapedTomato<T, false>;
export type RecordTomato<T extends Record<any, any>> = RecordShapedTomato<T, false>;

// Builders
export interface Breeds {
    string: AtomShapedTomato<string>;
    number: AtomShapedTomato<number>;
    boolean: AtomShapedTomato<boolean>;
    any: AtomShapedTomato<any>;
    array: <T extends AnyShapeTomato>(x: T) => ArrayTomato<T>;
    objectOf: <V>(val: V) => RecordTomato<Record<any, V>>;
    object: <V>(val: V) => ObjectTomato<V>;
}
