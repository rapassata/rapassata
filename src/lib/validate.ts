import { AnyShapeTomato, ArrayTomato, AtomShapedTomato, ObjectTomato, TomatoShape } from './tomatos';

enum FlowType {
    Transform = 'Transform',
    Validate = 'Validate',
}

interface TransformElement<I, O> {
    type: FlowType.Transform;
    transform: (x: I) => O;
}

interface ValidationElement<T> {
    type: FlowType.Validate;
    validate: (x: T) => Promise<Boolean> | boolean;
    message: string;
}

interface ValidationError {
    message: string;
    value: any;
}

export type FlowItem<T = any, O = any> = TransformElement<T, O> | ValidationElement<T>;

const baseNode = <B extends { shape: TomatoShape.Atom }, T>(base: B) => {
    const node: AtomShapedTomato<T> = {
        ...base,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node;
};

const atomNode = <T>() => {
    const node: AtomShapedTomato<T> = baseNode({
        shape: TomatoShape.Atom,
    });
    return node;
};

export const string: AtomShapedTomato<string> = (() => {
    const node = atomNode<string>();
    return node.validate(x => typeof x === 'string', 'Not a string');
})();

export const any: AtomShapedTomato<any> = (() => {
    return atomNode<any>();
})();

export const number: AtomShapedTomato<number> = (() => {
    const node = atomNode<number>();
    return node.validate(x => typeof x === 'number', 'Not a number');
})();

export const boolean: AtomShapedTomato<boolean> = (() => {
    const node = atomNode<boolean>();
    return node.validate(x => typeof x === 'boolean', 'Not a boolean');
})();

export const array = <T extends AnyShapeTomato>(item: T): ArrayTomato<T> => {
    const node: ArrayTomato<T> = {
        item,
        shape: TomatoShape.Array,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(x => Array.isArray(x), 'Not an array');
};

const isObject = (x: any) => typeof x === 'object' && x && x.constructor === Object;

export const object = <T>(structure: T): ObjectTomato<T> => {
    const node: ObjectTomato<T> = {
        structure,
        shape: TomatoShape.Object,
        flow: [],
        required: false,
        default: undefined,
        validate: (validate, message = '') => ({
            ...node,
            flow: [...node.flow, { message, validate, type: FlowType.Validate }],
        }),
        require: () => ({ ...node, required: true }),
        defaultTo: x => ({ ...node, default: x }),
    };
    return node.validate(isObject, 'Not an object');
};

const runFlow = (flow: FlowItem[], value: any) => {
    const errors: ValidationError[] = [];
    flow.reduce((value, flowItem) => {
        if (flowItem.type === FlowType.Validate && !flowItem.validate(value)) {
            errors.push({ value, message: flowItem.message });
        }
        return value;
    }, value);
    return {
        errors,
        value,
    };
};

export const validate = (schema: AnyShapeTomato, value: any): any => {
    const flowRes = runFlow(schema.flow, value);
    if (schema.shape === TomatoShape.Atom) {
        return flowRes;
    }
    if (schema.shape === TomatoShape.Array) {
        return {
            ...flowRes,
            children: (Array.isArray(value) ? value : []).map(v => validate(schema.item, v)),
        };
    }
    value = isObject(value) ? value : {};
    return {
        ...flowRes,
        structure: Object.keys(schema.structure).reduce(
            (res, key) => {
                res[key] = validate(schema.structure[key], value[key]);
                return res;
            },
            {} as any
        ),
    };
};
