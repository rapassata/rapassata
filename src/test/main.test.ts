import { string, validate, number, array, object, boolean } from 'lib/validate';

describe('Schemas', () => {
    test('string', () => {
        expect(string).toMatchSnapshot();
    });
    test('number', () => {
        expect(number).toMatchSnapshot();
    });
    test('boolean', () => {
        expect(boolean).toMatchSnapshot();
    });
    test('array', () => {
        expect(array(string)).toMatchSnapshot();
    });
    test('object', () => {
        expect(object(string)).toMatchSnapshot();
    });
});

describe('Validation (primitives)', () => {
    describe('string', () => {
        test('validate OK', () => {
            expect(validate(string, 'foo')).toMatchSnapshot();
        });
        test('validate null', () => {
            expect(validate(string, null)).toMatchSnapshot();
        });
    });
    describe('number', () => {
        test('validate OK', () => {
            expect(validate(number, 42)).toMatchSnapshot();
        });
        test('validate null', () => {
            expect(validate(number, null)).toMatchSnapshot();
        });
    });
    describe('boolean', () => {
        test('validate OK', () => {
            expect(validate(boolean, false)).toMatchSnapshot();
        });
        test('validate null', () => {
            expect(validate(boolean, null)).toMatchSnapshot();
        });
    });
    describe('array', () => {
        test('validate OK', () => {
            expect(validate(array(string), ['bar'])).toMatchSnapshot();
        });
        test('validate null', () => {
            expect(validate(array(string), null)).toMatchSnapshot();
        });
    });
    describe('object', () => {
        test('validate OK', () => {
            expect(validate(object({ a: string }), { a: 'x' })).toMatchSnapshot();
        });
        test('validate null', () => {
            expect(validate(object({ a: string }), null)).toMatchSnapshot();
        });
    });
});

describe('Validation (composite)', () => {
    describe('Schema 1', () => {
        const schema = object({
            user: object({
                age: number.defaultTo(18),
                address: object({
                    street: string.validate(x => x !== ''),
                    number: number,
                }),
                colors: array(string.defaultTo('red')).defaultTo([]),
                // ratedHobbies: objectOf(number).require(),
            }),
        });
        test('validate OK', () => {
            const value = {
                user: {
                    age: 12,
                    address: {
                        street: 'a',
                        number: 123,
                    },
                    colors: ['red'],
                },
            };
            expect(validate(schema, value)).toMatchSnapshot();
        });
        test('undefined', () => {
            expect(validate(schema, undefined)).toMatchSnapshot();
        });
    });
});
