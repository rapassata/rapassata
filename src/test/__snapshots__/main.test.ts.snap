// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schemas array 1`] = `
Object {
  "default": undefined,
  "defaultTo": [Function],
  "flow": Array [
    Object {
      "message": "Not an array",
      "type": "Validate",
      "validate": [Function],
    },
  ],
  "item": Object {
    "default": undefined,
    "defaultTo": [Function],
    "flow": Array [
      Object {
        "message": "Not a string",
        "type": "Validate",
        "validate": [Function],
      },
    ],
    "require": [Function],
    "required": false,
    "shape": "Atom",
    "validate": [Function],
  },
  "require": [Function],
  "required": false,
  "shape": "Array",
  "validate": [Function],
}
`;

exports[`Schemas boolean 1`] = `
Object {
  "default": undefined,
  "defaultTo": [Function],
  "flow": Array [
    Object {
      "message": "Not a boolean",
      "type": "Validate",
      "validate": [Function],
    },
  ],
  "require": [Function],
  "required": false,
  "shape": "Atom",
  "validate": [Function],
}
`;

exports[`Schemas number 1`] = `
Object {
  "default": undefined,
  "defaultTo": [Function],
  "flow": Array [
    Object {
      "message": "Not a number",
      "type": "Validate",
      "validate": [Function],
    },
  ],
  "require": [Function],
  "required": false,
  "shape": "Atom",
  "validate": [Function],
}
`;

exports[`Schemas object 1`] = `
Object {
  "default": undefined,
  "defaultTo": [Function],
  "flow": Array [
    Object {
      "message": "Not an object",
      "type": "Validate",
      "validate": [Function],
    },
  ],
  "require": [Function],
  "required": false,
  "shape": "Object",
  "structure": Object {
    "default": undefined,
    "defaultTo": [Function],
    "flow": Array [
      Object {
        "message": "Not a string",
        "type": "Validate",
        "validate": [Function],
      },
    ],
    "require": [Function],
    "required": false,
    "shape": "Atom",
    "validate": [Function],
  },
  "validate": [Function],
}
`;

exports[`Schemas string 1`] = `
Object {
  "default": undefined,
  "defaultTo": [Function],
  "flow": Array [
    Object {
      "message": "Not a string",
      "type": "Validate",
      "validate": [Function],
    },
  ],
  "require": [Function],
  "required": false,
  "shape": "Atom",
  "validate": [Function],
}
`;

exports[`Validation (composite) Schema 1 undefined 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Not an object",
      "value": undefined,
    },
  ],
  "structure": Object {
    "user": Object {
      "errors": Array [
        Object {
          "message": "Not an object",
          "value": undefined,
        },
      ],
      "structure": Object {
        "address": Object {
          "errors": Array [
            Object {
              "message": "Not an object",
              "value": undefined,
            },
          ],
          "structure": Object {
            "number": Object {
              "errors": Array [
                Object {
                  "message": "Not a number",
                  "value": undefined,
                },
              ],
              "value": undefined,
            },
            "street": Object {
              "errors": Array [],
              "value": undefined,
            },
          },
          "value": undefined,
        },
        "age": Object {
          "errors": Array [],
          "value": undefined,
        },
        "colors": Object {
          "children": Array [],
          "errors": Array [],
          "value": undefined,
        },
      },
      "value": undefined,
    },
  },
  "value": undefined,
}
`;

exports[`Validation (composite) Schema 1 validate OK 1`] = `
Object {
  "errors": Array [],
  "structure": Object {
    "user": Object {
      "errors": Array [],
      "structure": Object {
        "address": Object {
          "errors": Array [],
          "structure": Object {
            "number": Object {
              "errors": Array [],
              "value": 123,
            },
            "street": Object {
              "errors": Array [],
              "value": "a",
            },
          },
          "value": Object {
            "number": 123,
            "street": "a",
          },
        },
        "age": Object {
          "errors": Array [],
          "value": 12,
        },
        "colors": Object {
          "children": Array [
            Object {
              "errors": Array [],
              "value": "red",
            },
          ],
          "errors": Array [],
          "value": Array [
            "red",
          ],
        },
      },
      "value": Object {
        "address": Object {
          "number": 123,
          "street": "a",
        },
        "age": 12,
        "colors": Array [
          "red",
        ],
      },
    },
  },
  "value": Object {
    "user": Object {
      "address": Object {
        "number": 123,
        "street": "a",
      },
      "age": 12,
      "colors": Array [
        "red",
      ],
    },
  },
}
`;

exports[`Validation (primitives) array validate OK 1`] = `
Object {
  "children": Array [
    Object {
      "errors": Array [],
      "value": "bar",
    },
  ],
  "errors": Array [],
  "value": Array [
    "bar",
  ],
}
`;

exports[`Validation (primitives) array validate null 1`] = `
Object {
  "children": Array [],
  "errors": Array [
    Object {
      "message": "Not an array",
      "value": null,
    },
  ],
  "value": null,
}
`;

exports[`Validation (primitives) boolean validate OK 1`] = `
Object {
  "errors": Array [],
  "value": false,
}
`;

exports[`Validation (primitives) boolean validate null 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Not a boolean",
      "value": null,
    },
  ],
  "value": null,
}
`;

exports[`Validation (primitives) number validate OK 1`] = `
Object {
  "errors": Array [],
  "value": 42,
}
`;

exports[`Validation (primitives) number validate null 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Not a number",
      "value": null,
    },
  ],
  "value": null,
}
`;

exports[`Validation (primitives) object validate OK 1`] = `
Object {
  "errors": Array [],
  "structure": Object {
    "a": Object {
      "errors": Array [],
      "value": "x",
    },
  },
  "value": Object {
    "a": "x",
  },
}
`;

exports[`Validation (primitives) object validate null 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Not an object",
      "value": null,
    },
  ],
  "structure": Object {
    "a": Object {
      "errors": Array [
        Object {
          "message": "Not a string",
          "value": undefined,
        },
      ],
      "value": undefined,
    },
  },
  "value": null,
}
`;

exports[`Validation (primitives) string validate OK 1`] = `
Object {
  "errors": Array [],
  "value": "foo",
}
`;

exports[`Validation (primitives) string validate null 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Not a string",
      "value": null,
    },
  ],
  "value": null,
}
`;
